### PR TITLE
Add compact mode in Variant discriminators binary serialization

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Change binary serialization of Variant data type: add `compact` mode to avoid writing the same discriminator multiple times for granules with single variant or with only NULL values. Add MergeTree setting `use_compact_variant_discriminators_serialization` that is enabled by default. Note that Variant type is still experimental and backward-incompatible change in serialization is ok.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit e524f3817254488c2e43b3aa9ea350e236910493 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62774/e524f3817254488c2e43b3aa9ea350e236910493/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by Avogar** [comment]
Example of using compact vs basic mode for storing 100 millions of rows with single variant:
```
avogar@avogar-dev:~/tmp/server/data/default/test1/all_97_192_3$ cat count.txt
100000000
avogar@avogar-dev:~/tmp/server/data/default/test1/all_97_192_3$ ll
total 782312
drwxr-x---   2 avogar avogar      4096 Apr 18 13:44  ./
drwxr-x--- 119 avogar avogar      4096 Apr 18 13:44  ../
-rw-r-----   1 avogar avogar       530 Apr 18 13:44  checksums.txt
-rw-r-----   1 avogar avogar        92 Apr 18 13:44  columns.txt
-rw-r-----   1 avogar avogar         9 Apr 18 13:44  count.txt
-rw-r-----   1 avogar avogar        10 Apr 18 13:44  default_compression_codec.txt
-rw-r-----   1 avogar avogar 400466909 Apr 18 13:44  id.bin
-rw-r-----   1 avogar avogar     34114 Apr 18 13:44  id.cmrk2
-rw-r-----   1 avogar avogar         1 Apr 18 13:44  metadata_version.txt
-rw-r-----   1 avogar avogar      9699 Apr 18 13:44  primary.cidx
-rw-r-----   1 avogar avogar        95 Apr 18 13:44  serialization.json
-rw-r-----   1 avogar avogar         0 Apr 18 13:44 'v.Array(UInt64).bin'
-rw-r-----   1 avogar avogar       259 Apr 18 13:44 'v.Array(UInt64).cmrk2'
-rw-r-----   1 avogar avogar         0 Apr 18 13:44 'v.Array(UInt64).size0.bin'
-rw-r-----   1 avogar avogar       259 Apr 18 13:44 'v.Array(UInt64).size0.cmrk2'
-rw-r-----   1 avogar avogar         0 Apr 18 13:44  v.String.bin
-rw-r-----   1 avogar avogar       259 Apr 18 13:44  v.String.cmrk2
-rw-r-----   1 avogar avogar 400466909 Apr 18 13:44  v.UInt64.bin
-rw-r-----   1 avogar avogar     34114 Apr 18 13:44  v.UInt64.cmrk2
-rw-r-----   1 avogar avogar       445 Apr 18 13:44  v.variant_discr.bin                                   <----- Compact mode
-rw-r-----   1 avogar avogar     10520 Apr 18 13:44  v.variant_discr.cmrk2
avogar@avogar-dev:~/tmp/server/data/default/test1/all_97_192_3$ cd ../../test2/all_97_192_3
avogar@avogar-dev:~/tmp/server/data/default/test2/all_97_192_3$ ll
total 782756
drwxr-x---   2 avogar avogar      4096 Apr 18 13:44  ./
drwxr-x--- 119 avogar avogar      4096 Apr 18 13:44  ../
-rw-r-----   1 avogar avogar       532 Apr 18 13:44  checksums.txt
-rw-r-----   1 avogar avogar        92 Apr 18 13:44  columns.txt
-rw-r-----   1 avogar avogar         9 Apr 18 13:44  count.txt
-rw-r-----   1 avogar avogar        10 Apr 18 13:44  default_compression_codec.txt
-rw-r-----   1 avogar avogar 400466909 Apr 18 13:44  id.bin
-rw-r-----   1 avogar avogar     34114 Apr 18 13:44  id.cmrk2
-rw-r-----   1 avogar avogar         1 Apr 18 13:44  metadata_version.txt
-rw-r-----   1 avogar avogar      9699 Apr 18 13:44  primary.cidx
-rw-r-----   1 avogar avogar        95 Apr 18 13:44  serialization.json
-rw-r-----   1 avogar avogar         0 Apr 18 13:44 'v.Array(UInt64).bin'
-rw-r-----   1 avogar avogar       259 Apr 18 13:44 'v.Array(UInt64).cmrk2'
-rw-r-----   1 avogar avogar         0 Apr 18 13:44 'v.Array(UInt64).size0.bin'
-rw-r-----   1 avogar avogar       259 Apr 18 13:44 'v.Array(UInt64).size0.cmrk2'
-rw-r-----   1 avogar avogar         0 Apr 18 13:44  v.String.bin
-rw-r-----   1 avogar avogar       259 Apr 18 13:44  v.String.cmrk2
-rw-r-----   1 avogar avogar 400466909 Apr 18 13:44  v.UInt64.bin
-rw-r-----   1 avogar avogar     34114 Apr 18 13:44  v.UInt64.cmrk2
-rw-r-----   1 avogar avogar    445565 Apr 18 13:44  v.variant_discr.bin                                   <---- Basic mode
-rw-r-----   1 avogar avogar     14034 Apr 18 13:44  v.variant_discr.cmrk2
```

File: src/DataTypes/Serializations/SerializationVariant.h [hunk 1]
```diff
@@ -18,6 +25,17 @@ namespace DB
...
+/// Basic mode, when all discriminators are serialized as is row by row.
+/// Compact mode, when we avoid writing the same discriminators in granules when there is
+/// only one variant (or only NULLs) in the granule.
+/// In compact mode we serialize granules in the following format:
+/// <number of rows in granule><granule format><granule data>
+/// There are 2 different formats of granule - plain and compact.
+/// Plain format is used when there are different discriminators in this granule,
+/// in this format all discriminators are serialized as is row by row.
+/// Compact format is used when all discriminators are the same in this granule,
+/// in this case only this single discriminator is serialized.
```
Comment (Avogar) [inline src/DataTypes/Serializations/SerializationVariant.h:38]: This is the explanation of both modes

**Review by CurtizJ — APPROVED** [review]
LGTM

**Comment by clickgapai** [comment]
Hi @Avogar @CurtizJ — while reviewing this PR I found the following:

- **Coverage gap:** BASIC mode write path for all-NULL `Variant` column (`col.hasOnlyNulls()` branch) untested — https://github.com/ClickHouse/ClickHouse/pull/104082

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**